h2o-algos/src/test/java/hex/DKVManagerTest.java

```diff
@@ -0,0 +1,346 @@
+      assertNotNull(trainVal);
+      assertTrue(trainVal.isFrame());
+      assertFalse(trainVal.isDeleted());
+      assertNotNull(trainVal.get()); // The training frame should be there
+
+      // No shared vecs were deleted from the training frame as the sharedFrame was deleted.
+      for (Key k : trainingFrame.keys()) {
+        assertNotNull(DKV.get(k));
+      }
+
+
+    } finally {
+      Scope.exit();
+    }
+
+  }
+
+    /**
+     * Train a model from a frame. Retain the frame only and let the model be deleted. The frame should remain in DKV.
+     */
+    @Test
+    public void testdeleteModel_retainFrame() {
+        try {
+            Scope.enter();
+            Frame trainingFrame = parse_test_file("./smalldata/iris/iris_wheader.csv");
+            Scope.track(trainingFrame);
+
+          NaiveBayesModel.NaiveBayesParameters parms = new NaiveBayesModel.NaiveBayesParameters();
+            parms._train = trainingFrame._key;
+            parms._laplace = 0;
+            parms._response_column = trainingFrame._names[4];
+            parms._compute_metrics = false;
+
+            final Model model = new NaiveBayes(parms).trainModel().get();
+            Scope.track_generic(model);
+
+          DKVManager.retain(new Key[]{trainingFrame._key});
+
+          // Training frame of the model should be retained as well
+            final Value value = DKV.get(trainingFrame._key);
+            assertNotNull(value);
+            assertTrue(value.isFrame());
+            assertFalse(value.isDeleted());
+
+            assertNotNull(value.get()); // The training frame should be there
+            
+            assertNull(DKV.get(model._key)); // The model should be deleted
+
+        } finally {
+            Scope.exit();
+        }
+
+    }
+
+  // Retaining models & integration test with models is in h2o-algos subproject.
+  @Test
+  public void testRetainFrame() {
+    Frame frame = null;
+
+    try {
+      frame = TestUtil.parse_test_file("./smalldata/testng/airlines_train.csv");
+      DKVManager.retain(new Key[]{frame._key});
+      assertTrue(H2O.STORE.containsKey(frame._key));
+      assertNotNull(DKV.get(frame._key));
+
+      for (Vec vec : frame.vecs()) {
+        assertNotNull(vec._key);
+
+        for (int i = 0; i < vec.nChunks(); i++) {
+          assertNotNull(DKV.get(vec.chunkKey(i)));
+        }
+      }
+
+    } finally {
+      if (frame != null) frame.delete();
+    }
+  }
+
+  @Test
+  public void testRetainNothing() throws InterruptedException {
+    Frame frame = null;
+
+    try {
+      frame = TestUtil.parse_test_file("smalldata/testng/airlines_train.csv");
+      DKVManager.retain(new Key[]{});
+      assertNull(DKV.get(frame._key));
+
+    } finally {
+      if (frame != null) frame.delete();
+    }
+  }
+
+    private static void testRetainFrame(Frame trainingFrame) {
+      DKVManager.retain(new Key[]{trainingFrame._key});
+      assertTrue(H2O.STORE.containsKey(trainingFrame._key));
```

Comment:
IMHO not entirely correct, please use DKV.get instead

Reply:
Fixed in https://github.com/h2oai/h2o-3/pull/3577/commits/4fae893626b410e4c67fcdb31261cbb6982b038c, multiple occurances.